src/main/sensors/opticalflow.c

```diff
@@ -0,0 +1,200 @@
+        return false;
+    }
+
+    opticalflow.dev.init(&opticalflow.dev);
+    opticalflow.quality = OPTICALFLOW_NO_NEW_DATA;
+    opticalflow.rawFlowRates.X = 0;
+    opticalflow.rawFlowRates.Y = 0;
+    opticalflow.processedFlowRates.X = 0;
+    opticalflow.processedFlowRates.Y = 0;
+    opticalflow.lastValidResponseTimeMs = millis();
+
+    if (opticalflowConfig()->rotation != 0) {
+        cosRotAngle = cosf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
+        sinRotAngle = sinf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
+    }
+    //low pass filter
+    if (opticalflowConfig()->flow_lpf != 0) {
+        const float flowCutoffHz = (float)opticalflowConfig()->flow_lpf / 100.0f;
+        const float flowGain     = pt2FilterGain(flowCutoffHz, opticalflow.dev.delayMs / 1000.0f);
+
+        pt2FilterInit(&xFlowLpf, flowGain);
+        pt2FilterInit(&yFlowLpf, flowGain);
+    }
+    return true;
+}
+
+void opticalflowUpdate(void) {
+    if (opticalflow.dev.update) {
+        opticalflow.dev.update(&opticalflow.dev);
+    }
+}
+
+void opticalflowProcess(void) {
+    opticalflowData_t data = {0};
+    bool hasNewData = opticalflow.dev.read(&opticalflow.dev, &data);
+    
+    opticalflow.quality = data.quality;
+    data.devMinQualityThreshold = opticalflow.dev.minQualityThreshold;
+
+    if (hasNewData) {
+        opticalflow.lastValidResponseTimeMs = millis();
+
+        opticalflow.rawFlowRates = data.flowRate;
+        opticalflow.deltaTimeUs  = data.deltaTimeUs;
+        
+        applySensorRotation(&opticalflow.processedFlowRates, &opticalflow.rawFlowRates);
+
+        applyLPF(&opticalflow.processedFlowRates);
+
+        // DEBUG SECTION
+        DEBUG_SET(DEBUG_OPTICALFLOW, 0, opticalflow.quality);
+        DEBUG_SET(DEBUG_OPTICALFLOW, 1, opticalflow.rawFlowRates.X);
+        DEBUG_SET(DEBUG_OPTICALFLOW, 2, opticalflow.rawFlowRates.Y);
+        DEBUG_SET(DEBUG_OPTICALFLOW, 3, opticalflow.processedFlowRates.X);
+        DEBUG_SET(DEBUG_OPTICALFLOW, 4, opticalflow.processedFlowRates.Y);
+        DEBUG_SET(DEBUG_OPTICALFLOW, 5, (int32_t)opticalflow.deltaTimeUs);
+    }
+}
+
+static void applySensorRotation(opticalflowRates_t * dist, opticalflowRates_t * src) {
+    dist->X = (int32_t)(src->X * cosRotAngle - src->Y * sinRotAngle);
+    dist->Y = (int32_t)(src->X * sinRotAngle + src->Y * cosRotAngle);
+    
+    if (opticalflowConfig()->flipY) {
+        dist->Y = -dist->Y;
+    }
+}
+
+static void applyLPF(opticalflowRates_t * flowRates) {
+    if (opticalflowConfig()->flow_lpf == 0) {
+        return;
+    }    
+
+    flowRates->X = pt2FilterApply(&xFlowLpf, flowRates->X);
+    flowRates->Y = pt2FilterApply(&yFlowLpf, flowRates->Y);
+}
+
+opticalflow_t * getLatestFlowOpticalflowData(void) {
+    return &opticalflow;
+}
+
+bool isOpticalflowHealthy(void) {
+    return cmp32(millis(), opticalflow.lastValidResponseTimeMs) < OPTICALFLOW_HARDWARE_TIMEOUT_MS;
```

Comment:
(this will fail after about 25 days without OF data ;-) )

Reply:
only on drones flying for 25 days with a single charge

Reply:
Well ... Siemens had this problem in one version of Logo PLCs (with unsigned overflow, so about 50 days). It did cost me quite a lot ...

Reply:
I can imagine,
By the way, do we consider flying for days in Betaflight?

Reply:
A lot of other code will fail in that case, so don't worry about it. 
`us` overflow is more problematic, only 35minutes in  31bits. Code usually uses 0 as invalid value, it is quite unlikely that it will cause collision (but better fix would be to return 1 instead of 0 in `micros()`)